src/lib/replication/log_store/home_raft_log_store.cpp

```diff
@@ -0,0 +1,268 @@
+    m_log_store->foreach (to_store_lsn(start), [end, &out_vec](store_lsn_t cur, const log_buffer& entry) -> bool {
+        bool ret = (cur < to_store_lsn(end) - 1);
+        if (cur < to_store_lsn(end)) { out_vec->emplace_back(to_nuraft_log_entry(entry)); }
+        return ret;
+    });
+    return out_vec;
+}
+
+nuraft::ptr< nuraft::log_entry > HomeRaftLogStore::entry_at(ulong index) {
+    nuraft::ptr< nuraft::log_entry > nle;
+    try {
+        auto log_bytes = m_log_store->read_sync(to_store_lsn(index));
+        nle = to_nuraft_log_entry(log_bytes);
+    } catch (const std::exception& e) {
+        REPL_STORE_LOG(ERROR, "entry_at({}) index out_of_range", index);
+        throw e;
+    }
+    return nle;
+}
+
+ulong HomeRaftLogStore::term_at(ulong index) {
+    ulong term;
+    try {
+        auto log_bytes = m_log_store->read_sync(to_store_lsn(index));
+        term = extract_term(log_bytes);
+    } catch (const std::exception& e) {
+        REPL_STORE_LOG(ERROR, "term_at({}) index out_of_range", index);
+        throw e;
+    }
+    return term;
+}
+
+raft_buf_ptr_t HomeRaftLogStore::pack(ulong index, int32_t cnt) {
+    static constexpr size_t estimated_record_size = 128;
+    size_t estimated_size = cnt * estimated_record_size + sizeof(uint32_t);
+
+    //   << Format >>
+    // # records (N)        4 bytes
+    // +---
+    // | log length (X)     4 bytes
+    // | log data           X bytes
+    // +--- repeat N
+    raft_buf_ptr_t out_buf = nuraft::buffer::alloc(estimated_size);
+    out_buf->put(cnt);
+
+    int32_t remain_cnt = cnt;
+    m_log_store->foreach (
+        to_store_lsn(index),
+        [this, &out_buf, &remain_cnt]([[maybe_unused]] store_lsn_t cur, const log_buffer& entry) mutable -> bool {
+            if (remain_cnt-- > 0) {
+                size_t avail_size = out_buf->size() - out_buf->pos();
+                if (avail_size < entry.size()) {
+                    avail_size += std::max(out_buf->size() * 2, (size_t)entry.size());
+                    out_buf = nuraft::buffer::expand(*out_buf, avail_size);
+                }
+                REPL_STORE_LOG(TRACE, "packing lsn={} of size={}, avail_size in buffer={}", to_repl_lsn(cur),
+                               entry.size(), avail_size);
+                out_buf->put(entry.bytes(), entry.size());
+            }
+            return (remain_cnt > 0);
+        });
+    return out_buf;
+}
+
+void HomeRaftLogStore::apply_pack(ulong index, nuraft::buffer& pack) {
+    pack.pos(0);
+    auto num_entries = pack.get_int();
+
+    auto slot = next_slot();
+    if (index < slot) {
+        // We are asked to apply/insert data behind next slot, so we must rollback before index and then append
+        m_log_store->rollback_async(to_store_lsn(index) - 1, nullptr);
+    } else if (index > slot) {
+        // We are asked to apply/insert data after next slot, so we need to fill in with dummy entries upto the slot
+        // before append the entries
```

Comment:
Those dummy entries are not touched upon replay. There is no option of failing to apply the pack. We have only 2 options, apply or crash, but upon coming back it will possibly do the same thing. This is an error in the system and we simply log, work it around.